File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D15.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y17D15(input: String) : Day {
    private val seedA: Long
    private val seedB: Long

    init {
        val seeds = input
            .lineSequence()
            .map { line -> line.takeLastWhile { it.isDigit() }.toLong() }
            .toList()
        seedA = seeds.first()
        seedB = seeds.last()
    }

    private val factorA = 16807
    private val factorB = 48271

    /* The generators both work on the same principle. To create its next value, a generator will take the previous 
    value it produced, multiply it by a factor (generator A uses 16807; generator B uses 48271), and then keep the 
    remainder of dividing that resulting product by 2147483647. That final remainder is the value it produces next.
        
    To calculate each generator's first value, it instead uses a specific starting value as its "previous value" 
    (as listed in your puzzle input). */
    private fun generator(seed: Long, factor: Int, multiples: Int? = null): Sequence<Short> =
        generateSequence(seed) { (it * factor) % Int.MAX_VALUE }
            .drop(1) // drop the seed value
            .let { sequence -> // filter for part2 rules (skip for part1)   
                if (multiples == null) sequence else sequence.filter { it % multiples == 0L }
            }.map(Long::toShort) // only look at lowest 16 bits

    /**
     * Used for both parts. Zips the two generator sequences together to create one sequence that generates pairs
     * of values generated by the generators. Returns how often the pairs match given the number of comparisions.
     */
    fun solve(comparisons: Int, generatorA: Sequence<Short>, generatorB: Sequence<Short>): Int =
        (generatorA zip generatorB) // join the two Generator sequences together
            .take(comparisons) // number of times the judge compares the generated values
            .count { (aValue, bValue) -> aValue == bValue } // count how many of those values compared are equal

    override fun part1() = solve(
        40_000_000,
        generator(seedA, factorA),
        generator(seedB, factorB)
    )

    override fun part2() = solve(
        5_000_000,
        generator(seedA, factorA, 4),
        generator(seedB, factorB, 8)
    )
}

fun main() = Day.runDay(Y17D15::class)

//    Class creation: 16ms
//    Part 1: 594 (1300ms)
//    Part 2: 328 (962ms)
//    Total time: 2279ms

File: rust/advent/src/bin/2017/y17d15.rs
use std::iter::successors;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input = (usize, usize);
type Output = usize;

const FACTOR_A: usize = 16_807;
const FACTOR_B: usize = 48_271;
const MOD: usize = 2_147_483_647;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 15).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.get_numbers().collect_tuple().unwrap()
}

fn generator(seed: usize, factor: usize, multiples: Option<usize>) -> impl Iterator<Item = u16> {
    successors(Some(seed), move |&it| Some((it * factor) % MOD))
        .skip(1)
        .filter(move |&it| {
            if let Some(multiples) = multiples {
                it % multiples == 0
            } else {
                true
            }
        })
        .map(|it| it as u16)
}

fn solve(input: Input, comparisons: usize, multiples_a: Option<usize>, multiples_b: Option<usize>) -> Output {
    let (seed_a, seed_b) = input;
    generator(seed_a, FACTOR_A, multiples_a)
        .zip(generator(seed_b, FACTOR_B, multiples_b))
        .take(comparisons)
        .filter(|(a, b)| a == b)
        .count()
}

fn part1(input: Input) -> Output {
    solve(input, 40_000_000, None, None)
}

fn part2(input: Input) -> Output {
    solve(input, 5_000_000, Some(4), Some(8))
}

#[test]
fn default() {
    let input = get_input(17, 15).unwrap();
    let input = parse_input(&input);
    assert_eq!(594, part1(input));
    assert_eq!(328, part2(input));
}

// Input parsed (168μs)
// 1. 594 (132ms)
// 2. 328 (186ms)
// Total: 319ms
